### PR TITLE
Mongo DB: Massively simplify generated queries

### LINKED ISSUE #13097: Multiple metrics with count distinct with MongoDB returns the same result
### Describe the bug
Specifying multiple "metrics" using "Number of distinct values of ..." does not work with MongoDB and reports _incorrect_ results. This works as expected on Postgres, RedShift.

#### Query: 
Summary [ Distinct values of first_name, Distinct values of last_name ]

#### Postgres

**Data:**
| id | first_name | last_name |
|:--|:--|:--|
| 1 | Jon | Doe |
| 2 | Jon | Snow |
| 3 | Jane | Smith |

**Result:**
| Distinct values of first_name | Distinct values of last_name |
|:--|:--|
| 2 | 3 |

#### MongoDB

**Data:**

```json
[
  {
    "_id": 1,
    "first_name": "Jon",
    "last_name": "Doe"
  },
  {
    "_id": 2,
    "first_name": "Jon",
    "last_name": "Snow"
  },
  {
    "_id": 3,
    "first_name": "Jane",
    "last_name": "Smith"
  }
]
```

**Result:**
| Distinct values of first_name | Distinct values of last_name |
|:--|:--|
| 2 | 2 |

Reversing the query order to:
Summary [ Distinct values of first_name, Distinct values of last_name ]

...results in:
| Distinct values of last_name | Distinct values of first_name |
|:--|:--|
| 3 | 3 |

**Expected behavior**
Expected the results to match the result for both postgres and mongodb.

**Information about your Metabase Installation:**

- version: 0.35.3 (Also on 0.36.3)
- java.vm.version: 11.0.5+10
- os.name: Linux
- jdbc-driver: PostgreSQL JDBC Driver (42.2.8)

**Severity**
How severe an issue is this bug to you? Is this annoying, blocking some users, blocking an upgrade or blocking your usage of Metabase entirely?

- Requiring us to double-check all metrics that use MongoDB
- We're having to write custom aggregation queries, requiring dev-time and no way to self-service

### PR DESCRIPTION
**Pull request by camsaul**
- Eliminate use for `$project` stages entirely.
   - before `$match` stages
   - before `$group` stages
   - when using temporal bucketing (`:datetime-field`)
   - when using nested fields

This results in queries that are significantly simpler and an order of magnitude more efficient.
   
Implements #4216 
Implements #14838
Fixes #13097

### CODE REVIEW

File: docs/administration-guide/01-managing-databases.md [hunk 1]
```diff
@@ -15,7 +15,7 @@ Now you’ll see a list of your databases. To connect another database to Metaba
 - Google Analytics
 - [Google BigQuery](databases/bigquery.md)
 - H2
-- [MongoDB (version 3.4 or higher)](databases/mongodb.md)
+- [MongoDB (version 3.6 or higher)](databases/mongodb.md) <!-- MongoDB supported version is from https://www.mongodb.com/support-policy -->
```
Comment (camsaul) [inline docs/administration-guide/01-managing-databases.md:18]: 3.4 was EOL'ed over a year ago (4.4 is the current version), and we can really generate much more efficient queries using some 3.6+ features, so it's time to bump our min supported version

**Comment by rlotun** [comment]
Just curious - do you think this touches on this issue in any way: https://github.com/metabase/metabase/issues/13097?

**Comment by camsaul** [comment]
@rlotun I wrote a test for #13097 to check if it was fixed... it wasn't, but I fixed it. So it is now

**Comment by paoliniluis** [comment]
Hi @camsaul, will this solve this one: https://github.com/metabase/metabase/issues/10741?

**Comment by camsaul** [comment]
@paoliniluis yeah, #10741 is actually a duplicate of #4216

**Comment by dpsutton** [comment]
don't have much context to evaluate but nothing stands out to me as bad. so nice job!

**Review by dpsutton — APPROVED** [review]

**Comment by camsaul** [comment]
@robdaemon are you 👍 on me merging this? This is the PR that bumps the min required MongoDB version form 3.4 to 3.6

**Review by robdaemon — APPROVED** [review]

**Pull request merged**